cosmwasm/packages/accounting/src/contract.rs

```diff
@@ -0,0 +1,968 @@
+        };
+
+        commit_transfer(deps.as_mut(), tx.clone()).unwrap();
+
+        // Repeating the transfer should return an error and should not change the balances.
+        commit_transfer(deps.as_mut(), tx.clone())
+            .expect_err("successfully committed duplicate transfer");
+
+        let src = account::Key::new(
+            tx.key.emitter_chain(),
+            tx.data.token_chain,
+            tx.data.token_address,
+        );
+        assert_eq!(tx.data.amount, *query_balance(deps.as_ref(), src).unwrap());
+
+        let dst = account::Key::new(
+            tx.data.recipient_chain,
+            tx.data.token_chain,
+            tx.data.token_address,
+        );
+        assert_eq!(tx.data.amount, *query_balance(deps.as_ref(), dst).unwrap());
+
+        assert_eq!(tx.data, query_transfer(deps.as_ref(), tx.key).unwrap());
+    }
+
+    #[test]
+    fn round_trip_transfer() {
+        let mut deps = mock_dependencies();
+        let tx = Transfer {
+            key: transfer::Key::new(3, [1u8; 32].into(), 5),
+            data: transfer::Data {
+                amount: Uint256::from(400u128),
+                token_chain: 3,
+                token_address: [3u8; 32].into(),
+                recipient_chain: 9,
+            },
+        };
+
+        commit_transfer(deps.as_mut(), tx.clone()).unwrap();
+
+        let rx = Transfer {
+            key: transfer::Key::new(tx.data.recipient_chain, [6u8; 32].into(), 2),
+            data: transfer::Data {
+                amount: tx.data.amount,
+                token_chain: tx.data.token_chain,
+                token_address: tx.data.token_address,
+                recipient_chain: tx.key.emitter_chain(),
+            },
+        };
+
+        commit_transfer(deps.as_mut(), rx.clone()).unwrap();
+
+        let src = account::Key::new(
+            tx.key.emitter_chain(),
+            tx.data.token_chain,
+            tx.data.token_address,
+        );
+        assert_eq!(Balance::zero(), query_balance(deps.as_ref(), src).unwrap());
+
+        let dst = account::Key::new(
+            tx.data.recipient_chain,
+            tx.data.token_chain,
+            tx.data.token_address,
+        );
+        assert_eq!(Balance::zero(), query_balance(deps.as_ref(), dst).unwrap());
+
+        assert_eq!(tx.data, query_transfer(deps.as_ref(), tx.key).unwrap());
+        assert_eq!(rx.data, query_transfer(deps.as_ref(), rx.key).unwrap());
+    }
+
+    #[test]
+    fn missing_wrapped_account() {
+        let mut deps = mock_dependencies();
+        let tx = Transfer {
+            key: transfer::Key::new(9, [1u8; 32].into(), 5),
+            data: transfer::Data {
+                amount: Uint256::from(400u128),
+                token_chain: 3,
+                token_address: [3u8; 32].into(),
+                recipient_chain: 3,
+            },
+        };
+
+        commit_transfer(deps.as_mut(), tx)
+            .expect_err("successfully committed transfer with missing wrapped account");
```

Comment:
I think it would be better check for a specific error just to make sure it's the error we expect and not coincidentally hitting a different error.  My best stab at this with anyhow is this:

```rust
        let res = commit_transfer(deps.as_mut(), tx);
        let msg = res.err().unwrap().downcast_ref::<&str>().unwrap().clone();
        assert!(msg.contains("without an existing native account"));
```

Reply:
Done.  I added function-specific errors to make it easier to downcast and also added a couple of examples showing how to do that.